--- a/drivers/net/phy/at803x.c
+++ b/drivers/net/phy/at803x.c
@@ -319,6 +319,14 @@ static int at803x_config_init(struct phy_device *phydev)
 		if (pdata->disable_smarteee)
 			at803x_disable_smarteee(phydev);
 	}
+
+#ifdef CONFIG_OF_MDIO
+	if (phydev->mdio.dev.of_node &&
+		of_property_read_bool(phydev->mdio.dev.of_node,
+				      "at803x-disable-smarteee")) {
+		at803x_disable_smarteee(phydev);
+	}
+#endif
 
 	/* The RX and TX delay default is:
 	 *   after HW reset: RX delay enabled and TX delay disabled
